structured_data/2021_10_01_imbalanced_datasets/README.md

```diff
@@ -0,0 +1,130 @@
+  * Have a look at 'balanced accuracy' instead!
+* TPR, FPR and ROC AUC are a **slightly better choice**
+  * Note that the ROC curve is not a good visual illustration for highly imbalanced data, because the FPR does not drop drastically when the Total Real Negatives is huge. Take the whole confusion matrix into account!
+* F1 score and AUPR are the **best choice**
+  * The focus of the (Area Under) Precision-Recall curve on the minority class makes it an effective diagnostic for imbalanced classification models.
+
+## 2. Under-sample the majority class
+
+There are 2 stragies to consider:
+
+| Strategy                                                             | Method             |                                                                                                                |
+| -------------------------------------------------------------------- | ------------------ | -------------------------------------------------------------------------------------------------------------- |
+| Prototype selection (select samples from the original dataset)       | *Random*           | Randomly select prototypes.                                                                                    |
+|                                                                      | *NearMiss*         | Adding heuristics in order to filter prototypes in a smarter way. (→ 3 different implementations in imblearn!) |
+|                                                                      | *Tomek's links*    | Keeping every two samples from different classes that are the nearest neighbors of each other.                 |
+| Prototype generation (generate samples based on majority class data) | *ClusterCentroids* | Selecting the centroids of K-means                                                                             |
+
+Illustration of Tomek's links:
+
+![Illustration of Tomek's links](./figures/tomeks_links.png)
+
+Cluster centroids example with [imbalanced-learn](https://pypi.org/project/imbalanced-learn/) (previously `imblearn`):
+
+```python
+from collections import Counter
+
+from sklearn.datasets import make_classification
+from imblearn.under_sampling import ClusterCentroids
+
+X, y = make_classification(n_samples=5000, n_features=2, n_redundant=0,
+                          n_classes=3, n_clusters_per_class=1,
+                          weights=[0.01, 0.05, 0.94], random_state=0)
+print(sorted(Counter(y).items())) # [(0, 64), (1, 262), (2, 4674)]
+X_resampled, y_resampled = ClusterCentroids().fit_resample(X, y)
+print(sorted(Counter(y_resampled).items())) # [(0, 64), (1, 64), (2, 64)]
+```
+
+The following image shows the dataset before and after resampling:
+
+![Cluster centroids imblearn](./figures/clustercentroids.png)
+
+## 3. Over-sample the minority class
+
+This can be done in 3 different ways:
+
+* **Random sampling**: pretty straightforward.
```

Comment:
It is indeed pretty straightfoward, but might be good to explain a little bit more nonetheless. e.g. `duplication of already existing datapoints in a random order/fashion`